serverless/serverless
# Fix awsProvider.js : "Cannot use 'in' operator to search for '0'

## What did you implement:

Closes #5681.

## How did you implement it:

<del>Same handler warnings introduced in #5638 does not care about `${variable}` syntax.
This PR moves the warning code to after variable population.</del>

This is a bit more complex than my initial thought.

https://github.com/serverless/serverless/blob/07154d772f14da7e7e3c985f412099670a53b0a7/lib/Serverless.js#L149-L154

Function names, those to be warned regarding same handler, is populated in L149 of Serverless.js.
However, service name validation is performed in `Variables#prepopulateService` which is invoked before function name population.
So, if `functions` is not an array but a string (variable syntax), it causes the error.

So, same-handler warning should be moved to Service#validate which is invoked after function name population.

## How can we verify it:

1. `npm install -g exoego/serverless#fix-5681`

2. Run `sls deploy` against below. Verify deploy is successfull and same handler warnings are shown.

```yml
service:
  name: fix5681
provider:
  name: aws
  runtime: nodejs8.10 
functions: ${self:custom.lambdaFunctions}
custom:
  lambdaFunctions:
    functionA:
      handler: handler.hello
    functionB:
      handler: handler.hello
```

3. Run `sls deploy` against below. Verify deploy is aborted by stage name validation error.

```yml
service:
  name: fix5681
provider:
  name: aws
  stage: inv@lid # NOTE: Invalid stage name for API Gateway
  runtime: nodejs8.10 
functions: ${self:custom.lambdaFunctions}
custom:
  lambdaFunctions:
    functionA:
      handler: handler.hello
    functionB:
      handler: handler.hello
```
## Todos:


***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO
